cpp/src/io/parquet/reader_impl_helpers.hpp

```diff
@@ -0,0 +1,197 @@
+  }
+
+  [[nodiscard]] auto const& get_key_value_metadata() const { return keyval_maps; }
+
+  /**
+   * @brief Gets the concrete nesting depth of output cudf columns
+   *
+   * @param schema_index Schema index of the input column
+   *
+   * @return comma-separated index column names in quotes
+   */
+  [[nodiscard]] inline int get_output_nesting_depth(int schema_index) const
+  {
+    auto& pfm = per_file_metadata[0];
+    int depth = 0;
+
+    // walk upwards, skipping repeated fields
+    while (schema_index > 0) {
+      if (!pfm.schema[schema_index].is_stub()) { depth++; }
+      // schema of one-level encoding list doesn't contain nesting information, so we need to
+      // manually add an extra nesting level
+      if (pfm.schema[schema_index].is_one_level_list()) { depth++; }
+      schema_index = pfm.schema[schema_index].parent_idx;
+    }
+    return depth;
+  }
+
+  /**
+   * @brief Extracts the pandas "index_columns" section
+   *
+   * PANDAS adds its own metadata to the key_value section when writing out the
+   * dataframe to a file to aid in exact reconstruction. The JSON-formatted
+   * metadata contains the index column(s) and PANDA-specific datatypes.
+   *
+   * @return comma-separated index column names in quotes
+   */
+  [[nodiscard]] std::string get_pandas_index() const;
+
+  /**
+   * @brief Extracts the column name(s) used for the row indexes in a dataframe
+   *
+   * @param names List of column names to load, where index column name(s) will be added
+   */
+  [[nodiscard]] std::vector<std::string> get_pandas_index_names() const;
+
+  /**
+   * @brief Filters and reduces down to a selection of row groups
+   *
+   * The input `row_start` and `row_count` parameters will be recomputed and output as the valid
+   * values based on the input row group list.
+   *
+   * @param row_group_indices Lists of row groups to read, one per source
+   * @param row_start Starting row of the selection
+   * @param row_count Total number of rows selected
+   *
+   * @return A tuple of corrected row_start, row_count and list of row group indexes and its
+   *         starting row
+   */
+  [[nodiscard]] std::tuple<size_type, size_type, std::vector<row_group_info>> select_row_groups(
+    host_span<std::vector<size_type> const> row_group_indices,
+    size_type row_start,
+    size_type row_count) const;
+
+  /**
+   * @brief Filters and reduces down to a selection of columns
+   *
+   * @param use_names List of paths of column names to select; `nullopt` if user did not select
+   * columns to read
+   * @param include_index Whether to always include the PANDAS index column(s)
+   * @param strings_to_categorical Type conversion parameter
+   * @param timestamp_type_id Type conversion parameter
+   *
+   * @return input column information, output column information, list of output column schema
+   * indices
+   */
+  [[nodiscard]] std::
+    tuple<std::vector<input_column_info>, std::vector<column_buffer>, std::vector<size_type>>
```

Comment:
Consider returning a struct here; the type name is so long that auto-formatting is ugly :D

Reply:
Unfortunately we have to assign the results to existing variables through `std::tie` do a struct can't be extracted by that way:
```
std::tie(_input_columns, _output_buffers, _output_column_schemas) = _metadata->select_columns(...);
```